pkg/gadget-collection/gadgets/traceloop/gadget.go

```diff
@@ -52,138 +67,298 @@ some differences:
+
+		// The surrounding function can be called from any context.
+		// So, we need to manually patch the trace CRD to have our modifications be
+		// taken into account.
+		err = t.client.Status().Patch(context.TODO(), trace, patch)
+		if err != nil {
+			log.Errorf("Failed to patch trace %q output: %s", trace.Name, err)
+
+			return err
+		}
+
+		return nil
+	}
+
+	detachContainerFunc := func(container *containercollection.Container) {
+		mntNsID := container.Mntns
+		key := genKey(container)
+
+		traceUnique.Lock()
+		err := traceUnique.tracer.Detach(mntNsID)
+		traceUnique.Unlock()
+		if err != nil {
+			log.Errorf("failed to detach tracer: %s", err)
+
+			return
+		}
+
+		log.Debugf("tracer detached for %q (%d)", key, mntNsID)
+	}
+
+	containerEventCallback := func(event containercollection.PubSubEvent) {
+		switch event.Type {
+		case containercollection.EventTypeAddContainer:
+			attachContainerFunc(event.Container)
+		case containercollection.EventTypeRemoveContainer:
+			detachContainerFunc(event.Container)
+		}
+	}
+
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		var err error
+
+		traceUnique.tracer, err = tracelooptracer.NewTracer(t.helpers)
+		if err != nil {
+			traceUnique.Unlock()
+
+			trace.Status.OperationError = fmt.Sprintf("Failed to start traceloop tracer: %s", err)
+
+			return
+		}
+	}
+	traceUnique.users++
+	traceUnique.Unlock()
+
+	existingContainers := t.helpers.Subscribe(
+		genPubSubKey(trace.ObjectMeta.Namespace+"/"+trace.ObjectMeta.Name),
+		*gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter),
+		containerEventCallback,
+	)
+
+	for _, c := range existingContainers {
+		err := attachContainerFunc(c)
+		if err != nil {
+			log.Warnf("Warning: couldn't attach BPF program: %s", err)
+			break
+		}
+	}
 	t.started = true
 
 	trace.Status.State = gadgetv1alpha1.TraceStateStarted
 }
 
-func (t *Trace) stop() error {
-	if !t.started {
-		return errors.New("not started")
+func (t *Trace) Collect(trace *gadgetv1alpha1.Trace) {
+	if trace.Spec.OutputMode != gadgetv1alpha1.TraceOutputModeStream {
+		trace.Status.OperationError = fmt.Sprintf("\"collect\" operation can only be used with %q trace while %q was given", gadgetv1alpha1.TraceOutputModeStream, trace.Spec.OutputMode)
+
+		return
 	}
 
-	err := t.cmd.Process.Signal(syscall.SIGINT)
-	if err != nil {
-		return fmt.Errorf("failed to send SIGINT to process: %w", err)
+	traceUnique.Lock()
+	if traceUnique.tracer == nil {
+		traceUnique.Unlock()
+
+		trace.Status.OperationError = "Traceloop tracer is nil"
+
+		return
 	}
+	traceUnique.Unlock()
+
+	containerID := trace.Spec.Parameters["containerID"]
+	_, ok := t.containerIDs[containerID]
+	if !ok {
+		ids := make([]string, len(t.containerIDs))
+		i := 0
+		for id := range t.containerIDs {
+			ids[i] = id
+		}
 
-	timeout := time.After(2 * time.Second)
+		trace.Status.OperationError = fmt.Sprintf("%q is not a valid ID for this trace, valid IDs are: %v", containerID, strings.Join(ids, ","))
```

Comment:
The error message is malformed:
```
time="2022-11-16T12:10:30Z" level=info msg="Updating new status of trace \"gadget/traceloop\": state=Completed operationError=\"\\\"XX317b6937f7b79e5dfcdf0c986db039ced4eff55cdd45b537e1aa9b02a95909de\\\" is not a valid ID for this trace, valid IDs are: db2cdf35ea5ef76315f8eeb51a8f3d5e46142e0c03ebaae67669d3a767b603a4,,,,,,,\" operationWarning=\"\" output=<0 characters>"
```

```
status:
  operationError: '"XX317b6937f7b79e5dfcdf0c986db039ced4eff55cdd45b537e1aa9b02a95909de"
    is not a valid ID for this trace, valid IDs are: db2cdf35ea5ef76315f8eeb51a8f3d5e46142e0c03ebaae67669d3a767b603a4,,,,,,,'
```

This is because `i` is the constant `0` , so it only sets `ids[0]`.

I suggest the following code:
```
ids := []string{}
for id := range t.containerIDs {
    ids = append(ids, id)
}
```

This can be fixed in the next PR.